compiler/src/dotty/tools/backend/sjs/JUnitBootstrappers.scala

```diff
@@ -0,0 +1,283 @@
+        New(mydefn.TestMetadataType, List(name, ignored, reifiedAnnot))
+      }
+      JavaSeqLiteral(metadata, TypeTree(mydefn.TestMetadataType))
+    })
+  }
+
+  private def genInvokeTest(owner: ClassSymbol, testClass: ClassSymbol, tests: List[Symbol])(
+      implicit ctx: Context, mydefn: MyDefinitions): DefDef = {
+
+    val sym = ctx.newSymbol(owner, Names.invokeTest, Synthetic | Method,
+        MethodType(List(Names.instance, Names.name), List(defn.ObjectType, defn.StringType), mydefn.FutureType)).entered
+
+    DefDef(sym, { (paramRefss: List[List[Tree]]) =>
+      val List(List(instanceParamRef, nameParamRef)) = paramRefss
+      tests.foldRight[Tree] {
+        val tp = mydefn.NoSuchMethodExceptionType
+        val constr = tp.member(nme.CONSTRUCTOR).suchThat { c =>
+          c.info.paramInfoss.head.size == 1 &&
+          c.info.paramInfoss.head.head.isRef(defn.StringClass)
+        }.symbol.asTerm
+        Throw(New(tp, constr, nameParamRef :: Nil))
+      } { (test, next) =>
+        If(Literal(Constant(test.name.toString)).select(defn.Any_equals).appliedTo(nameParamRef),
+            genTestInvocation(testClass, test, instanceParamRef),
+            next)
+      }
+    })
+  }
+
+  private def genTestInvocation(testClass: ClassSymbol, testMethod: Symbol, instance: Tree)(
+      implicit ctx: Context, mydefn: MyDefinitions): Tree = {
+
+    def castInstance = instance.cast(testClass.typeRef)
+
+    val resultType = testMethod.info.resultType
+    if (resultType.isRef(defn.UnitClass)) {
+      val newSuccess = ref(mydefn.SuccessModule_apply).appliedTo(ref(defn.BoxedUnit_UNIT))
+      Block(
+          castInstance.select(testMethod).appliedToNone :: Nil,
+          ref(mydefn.FutureModule_successful).appliedTo(newSuccess)
+      )
+    } else if (resultType.isRef(mydefn.FutureClass)) {
+      castInstance.select(testMethod).appliedToNone
+    } else {
+      // We lie in the error message to not expose that we support async testing.
+      ctx.error("JUnit test must have Unit return type", testMethod.sourcePos)
+      EmptyTree
+    }
+  }
+
+  private def genNewInstance(owner: ClassSymbol, testClass: ClassSymbol)(implicit ctx: Context): DefDef = {
+    val sym = ctx.newSymbol(owner, Names.newInstance, Synthetic | Method,
+        MethodType(Nil, defn.ObjectType)).entered
+
+    DefDef(sym, New(testClass.typeRef, Nil))
+  }
+
+  private def castParam(param: Symbol, clazz: Symbol)(implicit ctx: Context): Tree =
+    ref(param).cast(clazz.typeRef)
+
+  private def annotatedMethods(owner: ClassSymbol, annot: Symbol)(implicit ctx: Context): List[Symbol] =
+    owner.info.decls.filter(m => m.is(Method) && m.hasAnnotation(annot))
+}
+
+private object JUnitBootstrappers {
+  private object ExtractAPI {
+    import xsbti.api
+
+    def createAndEmitSbtClassLike(bootstrapperClass: ClassSymbol)(implicit ctx: Context): Unit = {
+      // See computeClass() in ExtractAPI.scala
```

Comment:
zinc-apiinfo (that we cannot depend on) has APIUtil#emptyClassLike to do something similar: https://github.com/sbt/zinc/blob/c46643f3e68d7d4f270bf318e3f150f5a59c0aab/internal/zinc-apiinfo/src/main/scala/xsbt/api/APIUtil.scala#L95

I suggest copying a subset of that in a new file in https://github.com/lampepfl/dotty/tree/master/compiler/src/dotty/tools/dotc/sbt for use here and potentially in plugins.

Reply:
We could just have a single public method called `registerDummyClass` or something whose job it is to call the callback.

Reply:
Done.